internal/stores/contracts.go

```diff
@@ -356,6 +364,61 @@ func (c dbContract) convert() (rhpv2.Contract, error) {
 	}, nil
 }
 
+// AcquireContract acquires a contract assuming that the contract exists and
+// that it isn't locked right now. The returned bool indicates whether locking
+// the contract was successful.
+func (s *SQLStore) AcquireContract(fcid types.FileContractID, duration time.Duration) (types.FileContractRevision, bool, error) {
+	var contract dbContract
+	var locked bool
+
+	fcidGob := bytes.NewBuffer(nil)
+	if err := gob.NewEncoder(fcidGob).Encode(fcid); err != nil {
+		return types.FileContractRevision{}, false, err
+	}
+
+	err := s.db.Transaction(func(tx *gorm.DB) error {
+		// Get revision.
+		err := tx.Model(&dbContract{}).
+			Where("fcid", fcidGob.Bytes()).
+			Preload("Revision").
+			Take(&contract).
+			Error
+		if err != nil {
+			return err
+		}
+		// See if it is locked.
+		locked = time.Now().Before(contract.LockedUntil)
+		if locked {
+			return nil
+		}
+
+		// Update lock.
+		return tx.Model(&dbContract{}).
+			Where("fcid", fcidGob.Bytes()).
+			Update("locked_until", time.Now().Add(duration).UTC()).
+			Error
+	})
+	if err != nil {
+		return types.FileContractRevision{}, false, fmt.Errorf("failed to lock contract: %w", err)
+	}
+	if locked {
+		return types.FileContractRevision{}, false, nil
+	}
+	return contract.Revision.convert(fcid), true, nil
+}
+
+// ReleaseContract releases a contract by setting its locked_until field to 0.
+func (s *SQLStore) ReleaseContract(fcid types.FileContractID) error {
+	fcidGob := bytes.NewBuffer(nil)
+	if err := gob.NewEncoder(fcidGob).Encode(fcid); err != nil {
```

Comment:
I thought gorm could do this sort of thing automatically? Why do we need to explicitly encode here?